components/brave_new_tab_ui/components/default/rewards/index.tsx

```diff
@@ -0,0 +1,281 @@
+  enabledAds: boolean
+  enabledMain: boolean
+  balance: NewTab.RewardsBalance
+  grants: NewTab.GrantRecord[]
+  reports: Record<string, NewTab.RewardsReport>
+  walletCreated: boolean
+  walletCreating: boolean
+  walletCreateFailed: boolean
+  walletCorrupted: boolean
+  adsEstimatedEarnings: number
+  onlyAnonWallet?: boolean
+  onCreateWallet: () => void
+  onEnableAds: () => void
+  onEnableRewards: () => void
+  onDismissNotification: (id: string) => void
+}
+
+const enum AmountItemType {
+  ADS = 0,
+  TIPS = 1
+}
+
+class Rewards extends React.PureComponent<RewardsProps, {}> {
+  getTotalContributions = () => {
+    const currentTime = new Date()
+    const reportKey = `${currentTime.getFullYear()}_${currentTime.getMonth() + 1}`
+    const report: NewTab.RewardsReport = this.props.reports[reportKey]
+
+    if (report) {
+      return tipsTotal(report)
+    }
+
+    return '0.0'
+  }
+
+  getButtonText = (isAds: boolean = false) => {
+    if (isAds) {
+      return getLocale('turnOnAds')
+    }
+
+    const {
+      walletCreating,
+      walletCreateFailed,
+      walletCorrupted
+    } = this.props
+
+    if (walletCreateFailed || walletCorrupted) {
+      return getLocale('walletFailedButton')
+    }
+
+    if (walletCreating) {
+      return getLocale('rewardsWidgetTurningOn')
+    }
+
+    return getLocale('rewardsWidgetTurnOnRewards')
+  }
+
+  optInAction = (hasEnabled: boolean) => {
+    if (hasEnabled) {
+      this.props.onEnableRewards()
+    } else {
+      this.props.onCreateWallet()
+    }
+  }
+
+  renderPreOptIn = () => {
+    const {
+      enabledMain,
+      walletCreated
+    } = this.props
+
+    if (enabledMain && walletCreated) {
+      return null
+    }
+
+    const hasEnabled = !enabledMain && walletCreated
+
+    return (
+      <>
+        <PreOptInInfo>
+          <Title>
+            {
+              hasEnabled
+              ? getLocale('rewardsWidgetReEnableTitle')
+              : getLocale('rewardsWidgetEnableTitle')
+            }
+          </Title>
+          <SubTitle>
+            {
+              hasEnabled
+              ? getLocale('rewardsWidgetReEnableSubTitle')
+              : getLocale('rewardsWidgetEnableSubTitle')
+            }
+          </SubTitle>
+        </PreOptInInfo>
+        <PreOptInAction>
+          <TurnOnButton
+            data-test-id={'optInAction'}
+            onClick={this.optInAction.bind(this, hasEnabled)}
+          >
+            {this.getButtonText()}
+          </TurnOnButton>
+        </PreOptInAction>
+      </>
+    )
+  }
+
+  renderAmountItem = (type: AmountItemType) => {
+    const {
+      balance,
+      enabledAds,
+      onEnableAds,
+      adsEstimatedEarnings,
+      onlyAnonWallet
+    } = this.props
+
+    const rates = balance.rates || {}
+    const showEnableAds = type === AmountItemType.ADS && !enabledAds
+    const amount = type === AmountItemType.ADS
+      ? adsEstimatedEarnings.toFixed(1)
+      : this.getTotalContributions()
```

Comment:
`render()` should only do simple data rendering. This seems like complex calculations, especially BigNumber-related stuff. And it looks like the outcome is not dependent on all of the props of this component, but anything in render will run when any prop changes. These props here don't look high-fluctuating but I'd rather we don't introduce new anti-patterns (or copy existing ones from rewards). Options are to either:
- do it when only dependent actions are dispatched by calculating in the reducer and storing the result in state
- do it when only dependent properties change by using a library like reselect
- do it when only dependent properties change by using `componentDidUpdate` and seeing what changed

Reply:
this contribution total is now calculated in the reducer and stored in the state.